### PR TITLE
feat(query processor) Support Lambda Expressions

### PR DESCRIPTION
**Pull request by fpacifici**
Clickhouse supports lambda expressions and so should we.

This introduces a node in the ast that can be formatted into
```
arrayExists((x, y -> testFunc(x, y)), test)
```

### CODE REVIEW

File: snuba/clickhouse/formatter.py [hunk 1]
```diff
@@ -82,3 +84,11 @@ def visitCurriedFunctionCall(self, exp: CurriedFunctionCall) -> str:
         int_func = exp.internal_function.accept(self)
         ret = f"{int_func}{self.__visit_params(exp.parameters)}"
         return self.__alias(ret, exp.alias)
+
+    def visitVariable(self, exp: Variable) -> str:
+        return escape_identifier(exp.name)
+
+    def visitLambda(self, exp: Lambda) -> str:
+        variables = [escape_identifier(v) for v in exp.variables]
```
Comment (tkaemming) [inline snuba/clickhouse/formatter.py]: ~Doesn't this circumvent `visitVariable` by not calling `accept`?~

I initially misinterpreted this, the implementation makes sense to me now. This is an extreme nit, but I think it might be clearer to refer to these as "parameters" instead of "variables".

File: snuba/query/expressions.py [hunk 2]
```diff
@@ -214,3 +222,53 @@ def __iter__(self) -> Iterator[Expression]:
...
+
+@dataclass(frozen=True)
+class Lambda(Expression):
+    """
+    A lambda expression in the form (x,y,z -> transform(x,y,z))
+    """
+
+    # the bound variables in the expressions. These are intentionally not expressions
+    # since they are variable names and cannot have aliases
+    variables: Sequence[str]
```
Comment (tkaemming) [inline snuba/query/expressions.py]: Same nit re: `parameters`.

File: tests/clickhouse/test_formatter.py [hunk 3]
```diff
@@ -15,7 +17,6 @@
     (Literal(None, 123), "123",),  # INT literal
     (Literal(None, 123.321), "123.321",),  # FLOAT literal
     (Literal(None, None), "NULL",),  # NULL
-    (Literal(None, True), "true",),  # True
```
Comment (tkaemming) [inline tests/clickhouse/test_formatter.py]: Was this an intentional removal? This seems unrelated.

Comment (fpacifici) [reply tests/clickhouse/test_formatter.py]: No it is not. Actually I am not sure how I did that.

File: snuba/clickhouse/formatter.py (hunk 1 above)
Comment (fpacifici) [reply snuba/clickhouse/formatter.py]: Are you sure about this ? I used the term `variable` to follow the convention generally used when describing lambdas in languages that support them (which I believe comes from the formal definition in lambda calculus https://en.wikipedia.org/wiki/Lambda_calculus#Free_and_bound_variables). 
If this sounds confusing, what about `argument` so there is no naming conflict with the parameters of the FunctionCall class ?

Comment (tkaemming) [reply snuba/clickhouse/formatter.py]: > Are you sure about this ?

Not entirely. I'm not going to block on it by any means.

My understanding has generally been that `parameters` are the unbound variables of the function definition (incl. lambda functions), and `arguments` are the values of the bound variables. Python does refer to them as parameters in the language definition: https://docs.python.org/3/reference/expressions.html#lambda

> If this sounds confusing, what about argument so there is no naming conflict with the parameters of the FunctionCall class ?

I think this is more confusing, to be honest. In retrospect, if my understanding is correct, `FunctionCall.parameters` should have been `arguments`.

**Review by tkaemming — APPROVED** [review]
Leaving it up to you how/if you'd like to deal with `Lambda.variables` naming — it'll work either way.

**Comment by fpacifici** [comment]
@tkaemming what about this: 
- follow the python language definition
- r/variable/parameter
- r/Variable (class)/Argument

**Comment by tkaemming** [comment]
I think that generally makes sense, though I can't really say I've ever fully understood what the scope of an "argument" versus a "variable" is — whether it's only pertinent to the expression of the function call itself, or whether or not it makes sense to refer to the bound variable in the function body as an argument.

**Comment by fpacifici** [comment]
/though I can't really say I've ever fully understood what the scope of an "argument" versus a "variable" is /
I am about to send out a PR that makes the conditions_expr method (that today parses conditions) parametric. That will use this abstraction, there it should become clear.
